File: client-android/app/src/main/java/stasis/client_android/api/clients/MockServiceDiscoveryClient.kt
package stasis.client_android.api.clients

import stasis.client_android.lib.discovery.ServiceDiscoveryClient
import stasis.client_android.lib.discovery.ServiceDiscoveryRequest
import stasis.client_android.lib.discovery.ServiceDiscoveryResult
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Success

class MockServiceDiscoveryClient : ServiceDiscoveryClient {
    override val attributes: ServiceDiscoveryClient.Attributes =
        object : ServiceDiscoveryClient.Attributes {
            override fun asServiceDiscoveryRequest(isInitialRequest: Boolean): ServiceDiscoveryRequest =
                ServiceDiscoveryRequest(
                    isInitialRequest = isInitialRequest,
                    attributes = emptyMap()
                )
        }

    override suspend fun latest(isInitialRequest: Boolean): Try<ServiceDiscoveryResult> =
        Success(ServiceDiscoveryResult.KeepExisting)
}

File: core/src/test/scala/stasis/test/specs/unit/core/discovery/mocks/MockServiceDiscoveryClient.scala
package stasis.test.specs.unit.core.discovery.mocks

import java.util.concurrent.atomic.AtomicInteger

import scala.concurrent.Future

import stasis.core.discovery.ServiceDiscoveryClient
import stasis.core.discovery.ServiceDiscoveryResult

class MockServiceDiscoveryClient(
  initialDiscoveryResult: ServiceDiscoveryResult,
  nextDiscoveryResult: ServiceDiscoveryResult
) extends ServiceDiscoveryClient {
  private val counterRef: AtomicInteger = new AtomicInteger(0)

  override val attributes: ServiceDiscoveryClient.Attributes =
    MockServiceDiscoveryClient.TestAttributes(a = "b")

  override def latest(isInitialRequest: Boolean): Future[ServiceDiscoveryResult] = {
    val _ = counterRef.incrementAndGet()

    Future.successful(
      if (isInitialRequest) initialDiscoveryResult
      else nextDiscoveryResult
    )
  }

  def results: Int = counterRef.get()
}

object MockServiceDiscoveryClient {
  def apply(): MockServiceDiscoveryClient =
    new MockServiceDiscoveryClient(
      initialDiscoveryResult = ServiceDiscoveryResult.KeepExisting,
      nextDiscoveryResult = ServiceDiscoveryResult.KeepExisting
    )

  def apply(
    initialDiscoveryResult: ServiceDiscoveryResult,
    nextDiscoveryResult: ServiceDiscoveryResult
  ): MockServiceDiscoveryClient = new MockServiceDiscoveryClient(
    initialDiscoveryResult = initialDiscoveryResult,
    nextDiscoveryResult = nextDiscoveryResult
  )

  final case class TestAttributes(a: String) extends ServiceDiscoveryClient.Attributes
}
